modules/dmrpp_module/DmrppArray.cc

```diff
@@ -2021,51 +2129,51 @@ void DmrppArray::read_buffer_chunks() {
     if (get_chunk_count() < 2)
         throw BESInternalError(string("Expected chunks for variable ") + name(), __FILE__, __LINE__);
 
-    // Prepare buffer size.
-    unsigned long long max_buffer_end_position = 0;
+    // We need to pre-calculate the buffer_end_position for each buffer chunk to find the optimial buffer size.
+    unsigned long long buffer_offset = 0;
 
-    // For highly compressed chunks, we need to make sure the buffer_size is not too big because it may exceed the file
-    // size. For this variable we also need to find the maximum value of the end position of all the chunks. Here we try
-    // to loop through all the needed chunks for the constraint case.
-    bool first_needed_chunk = true;
-    unsigned long long first_needed_chunk_offset = 0;
-    unsigned long long first_needed_chunk_size = 0;
-    for (const auto &chunk : get_immutable_chunks()) {
+    // The maximum buffer size is limited to the current constrained domain.
+    unsigned long long max_buffer_size = bytes_per_element * this->get_maximum_constrained_buffer_nelmts();
+
+    // TODO: Sometimes the constraint size is too small and the constraint is across quite a few chunks. We may need
+    // to enlarge the max_buffer_size to 4K or even 1M. 
```

Comment:
4K Seems really small. Even 1M doesn't seem that large for an S3 transfer. Should we talk about this?

Reply:
We can.